### PR TITLE
Feature/csl703 no dht key

### PR DESCRIPTION
**Pull request by pva701**
* Now node is identified by `Peer` (pair ip:port) from outside.
* Store inside Kademlia map from `Peer` to `NodeId`

### CODE REVIEW

File: src/Network/Kademlia/Implementation.hs [hunk 1]
```diff
@@ -22,115 +23,123 @@ import           Control.Concurrent.Chan     (Chan, newChan, readChan)
...
+        modify $ \s -> s { known = [peer origin] }
+
+        -- Finish the lookup, recording which nodes returned the value
+        finish
+
+        -- Store the value in the closest peer that didn't return the
+        -- value
+        known <- gets known
+        polled <- gets polled
+        rest <- resolveNodeIds inst (polled \\ known)
```
Comment (georgeee) [inline src/Network/Kademlia/Implementation.hs]: I don't understand how  we can guarantee that nodes from `polled` are contained in buckets

File: src/Network/Kademlia/Implementation.hs [hunk 2]
```diff
@@ -141,56 +150,55 @@ data JoinResult
...
 
-          -- Continuing always means waiting for the next signal
-          continue = waitForReply finish checkSignal
+    -- Continuing always means waiting for the next signal
+    continue = waitForReply finish checkSignal
 
-          -- Send a FIND_NODE command, looking up your own id
-          sendS sendNode = liftIO ownId >>= flip sendSignal sendNode . FIND_NODE
+    -- Send a FIND_NODE command, looking up your own id
+    sendS toPeer = liftIO ownId >>= flip sendSignal toPeer . FIND_NODE
```
Comment (georgeee) [inline src/Network/Kademlia/Implementation.hs]: `toPeer` sounds like a function , let's leave  just `peer`

**Review by georgeee — CHANGES_REQUESTED** [review]

File: src/Network/Kademlia/Implementation.hs (hunk 1 above)
Comment (georgeee) [reply src/Network/Kademlia/Implementation.hs]: Or explain why we don't need his guarantee

Comment (pva701) [reply src/Network/Kademlia/Implementation.hs]: We can guarantee.

I. We add `peer` to `polled` when call `sendSignal`

IIa. when we call `sendSignal` at `startLookup`, we always calls `waitForReply` which does `insertNode` at the beginning

IIb. when we call `sendSignal` at `continueLookup` we every time call `continue`, but `continue=waitForReply` everywhere.

So we always call `waitForReply` which guarantee that a node will be inserted before any other actions.

Verify plz, and if it's true, I can add this comment to Kademlia.

Comment (georgeee) [reply src/Network/Kademlia/Implementation.hs]: But the fact we call `insertNode` doesn't imply node gets to any bucket. Appropriate bucket may be full for instance

File: src/Network/Kademlia/Implementation.hs [hunk 3]
```diff
@@ -266,65 +283,81 @@ startLookup cfg signalAction cancel onSignal = do
...
-                else do
-                    -- Insert the node into the tree, as it might be a new one or it
-                    -- would have to be refreshed
-                    liftIO . insertNode inst $ node
-
-                    case cmd of
-                      RETURN_NODES n nid _ -> do
+                removeFromEverywhere node >> continueIfMorePending
+            else do
+                when firstTime $ do
```
Comment (georgeee) [inline src/Network/Kademlia/Implementation.hs]: rename `firstTime` param to something like `withinJoin`

File: src/Network/Kademlia/ReplyQueue.hs [hunk 4]
```diff
@@ -78,8 +79,13 @@ data Reply i a = Answer (Signal i a)
 -- | The actual type representing a ReplyQueue
 data ReplyQueue i a = RQ {
       queue       :: (TVar [(ReplyRegistration i, Chan (Reply i a), ThreadId)])
+    -- ^ Queue of expected responses
     , timeoutChan :: Chan (Reply i a)
-    , defaultChan :: Chan (Reply i a)
+    -- ^ Channel for initial receiving of messages.
```
Comment (georgeee) [inline src/Network/Kademlia/ReplyQueue.hs]: What about renaming it to `dispatchChan`?

**Review by georgeee — APPROVED** [review]

**Pull request merged**